### ISSUE TITLE
docs: redesign event reference

### ISSUE THREAD

**Issue by onerandomusername** [DisnakeDev/disnake#460]
### Summary

seperate event reference into multiple types and sections

### What is the feature request for?

The documentation

### The Problem

There's too many events in a disorganized disarray to find a specific one very quicky.

### The Ideal Solution

There's a lot of events and events are important to disnake. Each type of event should have a section label and be listed in the sidebar.

**Comment by shiftinv** [comment]
somewhat related: #392

**Comment by ChristopherJHart** [comment]
I would be happy to take this on!

From running through the current events in master, I think we could break up events into the following sections:

## Client Events

* on_connect
* on_shard_connect
* on_disconnect
* on_shard_disconnect
* on_ready
* on_shard_ready
* on_resumed
* on_shard_resumed
* on_error
* on_socket_event_type
* on_socket_raw_receive
* on_socket_raw_send

## Chat Message Events

* on_typing
* on_raw_typing
* on_message
* on_message_delete
* on_raw_message_delete
* on_raw_bulk_message_delete
* on_message_edit
* on_raw_message_edit
* on_reaction_add
* on_raw_reaction_add
* on_reaction_remove
* on_raw_reaction_remove
* on_reaction_clear
* on_raw_reaction_clear
* on_reaction_clear_emoji
* on_raw_reaction_clear_emoji

## Interaction Events

* on_application_command_permissions_update
* on_interaction
* on_message_interaction
* on_button_click
* on_dropdown
* on_application_command
* on_application_command_autocomplete
* on_modal_submit

## Channel/Thread Events

* on_private_channel_update
* on_private_channel_pins_update
* on_guild_channel_delete
* on_guild_channel_update
* on_guild_channel_pins_update
* on_thread_join
* on_thread_remove
* on_thread_create
* on_thread_delete
* on_raw_thread_delete
* on_thread_member_join
* on_raw_thread_member_remove
* on_thread_update
* on_raw_thread_update
* on_group_join

## Guild Events

* on_guild_integrations_update
* on_integration_create
* on_integration_update
* on_raw_integration_delete
* on_webhooks_update
* on_member_join
* on_member_update
* on_presence_update
* on_user_update
* on_guild_join
* on_guild_remove
* on_guild_update
* on_guild_role_create
* on_guild_role_update
* on_guild_emojis_update
* on_guild_stickers_update
* on_guild_available
* on_guild_scheduled_event_create
* on_guild_scheduled_event_update
* on_guild_scheduled_event_subscribe
* on_raw_guild_scheduled_event_subscribe
* on_stage_instance_create
* on_stage_instance_update
* on_invite_create
* on_invite_delete
* on_member_ban
* on_member_unban

## Voice Events

* on_voice_state_update

I can envision some of these sections being broken up further (e.g. "Guild Integration Events", "Guild User/Member Events", etc.) too. I think this would be the first time in the documentations a third-level header/section is used, so if y'all are okay with that, I can implement it.

If y'all are okay with this, let me know and I'll get a PR open!

**Comment by bast0006** [comment]
Some things I'd suggest:
- It's good to repeat "events" for toplevel, it's bad for anything below that (avoid the "BobsEventManagerEventChatEvent" problem). 
- "Chat Message Events": try just "Message Events"? "Chat" is superfluous
- Contrary to shortening, I personally don't like the "Channel/Thread" bit, the / is visually not great. Perhaps "Channel and Thread"? This may just be a nit, I don't think "and" is much better

And finally, but with the most work and most importantly, some work should go into ordering the docs entries here. They're currently not.. visibly sorted by anything that makes sense (for example, the first under "Chat Message Events" is on_typing). 

Good organization would be semi-alphabetical, or by some kind of priority

on_reaction_add
on_reaction_remove
on_reaction_clear
on_raw_reaction_add
on_raw_reaction_remove
on_raw_reaction_clear

TLDR being alphabetical by type, then add/edit/remove/bulk/raw. Once this is consistent someone scrolling for an entry won't be forced to search needle-haystack style, and can instead spot where they are and crucially, if they've gone too far, whereas now it's not clear where different items are and I have to scroll the entire client events section looking for on_message_edit, for example. 

It'd be even better if there could be a more logical ordering overall than alphabetical, but that's harder

**Comment by ChristopherJHart** [comment]
I appreciate your suggestions! The latest push to PR #636:

* De-duplicates use of the word "events".
* Renamed `Chat Message Events` to just `Messages`.
* Modified section names to be plural (e.g. `Messages` instead of `Message`, `Interactions` instead of `Interaction`, etc.), since we're referencing more than one potential event.

I personally am not 100% sold on `Channels and Threads` instead of `Channels/Threads`. From my perspective, the two are equivalent (with the added bonus that `Channels/Threads` is shorter and gets the point across in less characters). I'm a native English speaker though, so maybe a non-native speaker would have a better perspective on the matter!

As far as organizing the events go, I do like your idea of alphabetizing, then sorting by CRUD-esque methods (I like your proposal of add/edit/remove/bulk/raw). Something I was considering is whether events should be broken up into non-raw and raw events, with all non-raw events in each section positioned above raw events. My thinking is that the *average* user is unlikely to care about most of the raw events and will likely favor the non-raw events.

To speak practically, I think you're proposing something like this:

```
on_message
on_message_edit
on_message_delete
on_bulk_message_delete
on_raw_message_delete
on_raw_message_edit
on_raw_bulk_message_delete
on_reaction_add
on_reaction_remove
on_reaction_clear
on_raw_reaction_add
on_raw_reaction_remove
on_raw_reaction_clear
```

What I'm proposing would be something like this:

```
on_message
on_message_edit
on_message_delete
on_bulk_message_delete
on_reaction_add
on_reaction_remove
on_reaction_clear
on_raw_message_delete
on_raw_message_edit
on_raw_bulk_message_delete
on_raw_reaction_add
on_raw_reaction_remove
on_raw_reaction_clear
```

Thoughts?

**Comment by onerandomusername** [comment]
I think raw events should stay ordered how they currently are, after their full counterpart. 
The gist of the raw events is that they are *always* ran, even when the cache does not include all objects. This means that on_message_delete may not fire, when on_raw_message_delete will. Most users will want to have the raw event so their code is ran even when the cache is empty or missing.

**Comment by ChristopherJHart** [comment]
The latest commit to PR #636 has been updated accordingly.
